Source/ASDisplayTree.mm

```diff
@@ -0,0 +1,278 @@
+  auto array = [self l_mutableSubnodesOf:parent];
+  if (location == ASTreeInsertReplace) {
+    array[index] = node;
+    [self l_createTreeFor:otherNode];
+  } else {
+    [array insertObject:node atIndex:index];
+  }
+  
+  [previousTree unlock];
+  [self releaseLocksForSelfAnd:node :otherNode];
+}
+
+- (void)remove:(ASDisplayNode *)node
+{
+  [self acquireLocksForSelfAnd:node :nil];
+  
+  // If the node is a root node, or if it's part of some other
+  // tree, then we have a race. Let it slide and ignore the call.
+  if (node->_supernode == nil || node->_tree != self) {
+    [self releaseLocksForSelfAnd:node :nil];
+    return;
+  }
+  
+  auto index = [self l_indexOf:node];
+  auto parent = [self supernodeOf:node];
+  [[self l_mutableSubnodesOf:parent] removeObjectAtIndex:index];
+  [self l_createTreeFor:node];
+  [self releaseLocksForSelfAnd:node :nil];
+}
+
+#pragma mark - Enumerating
+
+- (id<NSFastEnumeration>)l_enumeratingSubnodesOf:(ASDisplayNode *)node
+{
+  ASDisplayTreeAssertLocked();
+  return self;
+}
+
+- (id<NSFastEnumeration>)l_enumeratingUpwardFrom:(ASDisplayNode *)startNode
+{
+  ASDisplayTreeAssertLocked();
+  return self;
+}
+
+- (id<NSFastEnumeration>)l_enumeratingDownwardFrom:(ASDisplayNode *)startNode with:(ASTreeEnumerationOptions)options
+{
+  ASDisplayTreeAssertLocked();
+  return self;
+}
+
+- (id<NSFastEnumeration>)l_enumeratingUpwardFromParentOf:(ASDisplayNode *)startNode
+{
+  ASDisplayTreeAssertLocked();
+  return self;
+}
+
+#pragma mark - Treewide data
+
+- (ASPrimitiveTraitCollection)traitCollection
+{
+  return ASLockedSelf(_traitCollection);
+}
+
+- (void)setTraitCollection:(ASPrimitiveTraitCollection)traitCollection
+{
+  ASLockedSelf(_traitCollection = traitCollection);
+}
+
+#pragma mark - Internal
+
+- (NSMutableArray *)l_mutableSubnodesOf:(ASDisplayNode *)node
+{
+  ASDisplayTreeAssertLocked();
+  return node->_subnodes;
+}
+
+/*
+ * Convert a relative insertion point to an absolute point and a parent node.
+ */
+- (NSInteger)l_resolveInsertionPointFor:(ASTreeInsertionLocation)location relativeTo:(ASDisplayNode *)otherNode with:(NSInteger)index gettingParent:(ASDisplayNode **)parentPtr
+{
+  ASDisplayTreeAssertLocked();
+  NSParameterAssert(parentPtr);
+  switch (location) {
+    case ASTreeInsertAtIndex:
+      *parentPtr = otherNode;
+      return index;
+    case ASTreeInsertAtEnd:
+      *parentPtr = otherNode;
+      return [self l_mutableSubnodesOf:otherNode].count;
+    case ASTreeInsertAbove:
+      *parentPtr = [self supernodeOf:otherNode];
+      return [self l_indexOf:otherNode] + 1;
+    case ASTreeInsertBelow:
+    case ASTreeInsertReplace:
+      *parentPtr = [self supernodeOf:otherNode];
+      return [self l_indexOf:otherNode];
+  }
+}
+
+- (void)l_createTreeFor:(ASDisplayNode *)detachedNode
+{
+  ASDisplayTreeAssertLocked();
+}
+
+// Acquire self, and node instancelock, and other node instancelock
+- (void)acquireLocksForSelfAnd:(ASDisplayNode *)node :(ASDisplayNode *)otherNode
```

Comment:
lol :: in action.

Does this method make any assumptions or guarantees about which trees the node and otherNode are in? Is one of them guaranteed to be in the "self" tree, or not necessarily? Probably worth stating assumptions in the comment, or a clearer naming of the arguments.